stdlib/public/core/Filter.swift.gyb

```diff
@@ -295,6 +326,30 @@ public struct ${Self}<
     } while index != _base.endIndex && !_predicate(_base[index])
   }
 
+%   if Traversal == 'Bidirectional':
+  /// Returns the previous `index` of an element that `self._predicate` matches
+  /// otherwise `self._base.startIndex`
+  @inline(__always)
+  internal func _previousFiltered(_ index: Base.Index) -> Base.Index {
+    var index = index
+    _previousFilteredInPlace(&index)
+    return index
+  }
+
+  /// Returns the first index `i` preceding `index` where either
+  /// `_predicate(base[i]) == true` or `i == _base.startIndex`.
+  ///
+  /// - Precondition: `index != startIndex`
+  /// - Postcondition: `i < index`
+  @inline(__always)
+  internal func _previousFilteredInPlace(_ index: inout Base.Index) {
+    _precondition(index != _base.startIndex, "can't retreat before startIndex")
+    repeat {
+      _base.formIndex(before: &index)
+    } while index != _base.startIndex && !_predicate(_base[index])
```

Comment:
If the index is invalid, we only guarantee memory safety.  The program will likely trap because we will call `_base.formIndex(before: &index)` on the `startIndex`, or it might just continue looping forever if the base collection does not need range checks in `formIndex(before:)` to guarantee memory safety.

But the thing with invalid indices is that we can't detect them, so we don't care.  Operations that can compromise memory safety should perform such checks, but they can be located at a big distance from the point where the invalid index was first introduced.

As a matter of QoI, we try to provide redundant range checks to catch issues earlier, but if costs us `O(n)` -- because we have to check on each iteration whether we have reached `startIndex`, then it might not be worth it.